SIP-016 Metadata for Tokens (#44)
stacksgov/sips: sips/sip-x/sip-x-nft-metadata.md

# Preamble

SIP Number: X

Title: Schema Definition for Metadata for Non-Fungible Tokens

Author: Friedger Müffke (mail@friedger.de)

Consideration: Technical

Type: Standard

Status: Draft

Created: 20 October 2021

License: CC0-1.0

Sign-off:

# Abstract

> Non-fungible tokens or NFTs for short are digital assets registered on blockchain with unique identifiers and properties that distinguish them from each other. SIP-009 defines the trait how ownership of an NFT is managed. This SIP aims to provide a flexible standard to attach meta data to NFTs, like descriptions or urls to digital files.

Reviewer: ```suggestion
Non-fungible tokens or NFTs for short are digital assets registered on blockchain with unique identifiers and properties that distinguish them from each other. SIP-009 defines the trait for how ownership of an NFT is managed. This SIP aims to provide a flexible standard to attach metadata to NFTs, like descriptions or urls to digital files.
```

# Specification

Every SIP-X compliant smart contract in Stacks blockchain must be SIP-009 compliant and must meet the following requirements for the return value of function `get-token-uri`:

## Return Value of `get-token-uri`

The return value must be a `some` value if and only if the provided parameter `id` is the key of an NFT that was minted and not burnt, otherwise the value must be `none`.

For minted and not burnt NFTs, the inner value of the return value must be a string representing a resolvable URI. For string containing `{id}`, the `{id}` part must be replaced by the id in decimal format given in the function call.

The resolved data must be a JSON blob.

## JSON scheme of Metadata

The JSON blob resolved through the token uri must follow the following JSON schema. If the string `{id}` exists in any JSON value, it MUST be replaced with the actual token id in decimal format, by all client software that follows this standard.

> ```
> {
>     "title": "Token Metadata",

Reviewer: ```suggestion
```json
{
    "$schema": "http://json-schema.org/draft-07/schema#",
    "title": "Token Metadata",
```

    "type": "object",

>     "required": [],

Reviewer: No properties are required? How about additional properties?

    "properties": {
        "name": {
            "type": "string",

>             "description": "Identifies the asset to which this token represents"

Reviewer: Should there be a max length described?

                }
            }
        },
        "properties": {
            "type": "object",
            "description": "Arbitrary properties. Values may be strings, numbers, object or arrays."
        },
        "localization": {
            "type": "object",
            "required": ["uri", "default", "locales"],
            "properties": {
                "uri": {
                    "type": "string",
                    "description": "The URI pattern to fetch localized data from. This URI should contain the substring `{locale}` which will be replaced with the appropriate locale value before sending the request."
                },
                "default": {
                    "type": "string",
                    "description": "The locale of the default data within the base JSON"
                },
                "locales": {
                    "type": "array",
                    "description": "The list of locales for which data is available. These locales should conform to those defined in the Unicode Common Locale Data Repository (http://cldr.unicode.org/)."
                }
            }
        }
    }
}
```

### Examples

# Using NFTs in applications

> Before presenting metadata to users, application developers should verify whether the metadata is compliant with the applications guidelines.

Reviewer: ```suggestion
Before presenting metadata to users, application developers should verify whether the metadata is compliant with the application's guidelines.
```

Author: Friedger Müffke (mail@friedger.de)

Consideration: Technical

Type: Standard

Status: Draft

Created: 7 November 2021

License: CC0-1.0

Sign-off:

# Abstract

Non-fungible tokens or NFTs for short are digital assets registered on blockchain with unique identifiers and properties that distinguish them from each other. SIP-009 defines the trait for how ownership of an NFT is managed. This SIP aims to provide a flexible standard to attach metadata to NFTs, like descriptions or urls to digital files.

# License and Copyright

This SIP is made available under the terms of the Creative Commons CC0 1.0 Universal license, available at https://creativecommons.org/publicdomain/zero/1.0/
This SIP’s copyright is held by the Stacks Open Internet Foundation.

# Introduction

Tokens are digital assets registered on blockchain through a smart contract. A non-fungible token (NFT) is a token that is globally unique and can be identified through its unique identifier. In blockchains with smart contracts, including the Stacks blockchain, developers and users can use smart contracts to register and interact with non-fungible tokens.

> Some use cases of NFTs are name registration, digital art, certification, media and enternainment, real-estate. They all require that users associate certain content with an NFT. In general, it is helpful for the users to have a name, sound, image that represents this content.

Reviewer 2: nit: "entertainment"

Author: addressed in 661ba2a073b48aa4a45e30251182adff9a2f4057

# Specification

Every SIP-X compliant smart contract in Stacks blockchain must implement one or more functions that return a resolvable/retrievable uri referencing metadata. The metadata provide information for displaying a digital asset to users. This type of function is named "metadata uri functions".

Appendix A contains a list of trait functions that must meet the following requirements for the return value. The appendix can be extended without changing the ratification status of this SIP.

## Return Value of Metadata URI Functions

> The return value must be a `some` value if and only if the metadata reference an existing token, otherwise the value must be `none`. Appendix A specifies the exact meaning of "existingt" for each function.

Reviewer 2: nit: "existing"

Author: addressed in 661ba2a073b48aa4a45e30251182adff9a2f4057

For existing tokens, the inner value of the return value must be a string representing a resolvable URI. 

If a metadata uri function expects a parameter of type `uint` that identifies a token and the resulting strings contains `{id}`, then the `{id}` part must be replaced by the identifier in decimal format given in the function call.

The resolved data of the URI must be a JSON blob.

## JSON scheme of Metadata

The JSON blob resolved through the uri must follow the following JSON schema. 

> If metadata were retrieved by a function call containing a token identifier and the string `{id}` exists in any JSON value, it MUST be replaced with the actual token id in decimal format, by all client software that follows this standard.

Reviewer 2: I like this part :)

Author: :+1:

                },
                "default": {
                    "type": "string",
                    "description": "The locale of the default data within the base JSON"
                },
                "locales": {
                    "type": "array",
                    "description": "The list of locales for which data is available. These locales should conform to those defined in the Unicode Common Locale Data Repository (http://cldr.unicode.org/)."
                }
            }
        }
    }
}
```

The lengths of string values is not restricted. Nowadays, clients should be smart enough to deal with values of different lengths.

### Example
token101.json
```
{
  "version": 1,
  "name": "Foo #101",
  "image": "ipfs://somerandomecid",
  "attributes": [
     { 
      "trait_type": "hair",      
      "value": "red",
    },
    {
      "trait_type": "strength",
      "display_type": "number",
      "value": 99,
    },    
  ],
  "properties": {

>     { 

Reviewer 2: is this supposed to be nested or instead:
```
"properties": {
    "collection":  "Foo Collection",
    "collection_size":  "10000" 
}
```

Author: You are correct. Addressed in 661ba2a073b48aa4a45e30251182adff9a2f4057

| SIP-013 sip013-semi-fungible-token-trait.get-token-uri | token must be minted and not burnt, no requirements on the number of fungible part of the token|      |  1st |

# Appendix B

Attribute types

| Type | Display types | Additional Properties | Comment |
|------|---------------|-----------------------|---------|
| Numeric | `number`, `boost_percentage`, `boost_number` | `max_value`| |
| Date    | `date`     |                       | As unix timestamp in UTC | 
| String  | empty      |                       |         |

# Appendic C

Common Properties with predefined types.

| Name | Type | Description |
|------|------|-------------|
| `collection` | `string` | collection name the token belongs to. See also Appendix A. |
| `decimals` | `integer` | number of decimals. See also Appendix A. |
| `id` | `integer` | identifier for NFTs. See also Appendix A. |
| `created` | `integer` | creation date of the token in unix timestamp | 
| `symbol`  | `string` | token symbol |
| `total_supply`| `integer` | number of total supply, e.g. minted tokens |
| `ip_document_uri` | `string` | link to document about intellectual property (IP) rights |

> | `external_url` | `string` | url that will view the token on an external site |

Reviewer 3: another requested field is the `animation_url` which is often for a plain file type, but it can also support a URL to a full html page which is usually embedded in an iframe:

https://docs.opensea.io/docs/metadata-standards

Animation_url also supports HTML pages, allowing you to build rich experiences and interactive NFTs using JavaScript canvas, WebGL, and more. Scripts and relative paths within the HTML page are now supported. However, access to browser extensions is not supported.

Author: Has been added

For existing tokens, the inner value of the return value must be a string representing a resolvable URI.

The schema of the resolvable URI is not specified and should be a well-known schema like `https`, `ar`, `ipfs`, `sia`. A `data` URI is also valid, however, the length is limited by this SIP.

If a metadata uri function expects a parameter of type `uint` that identifies a token and the resulting strings contains `{id}`, then the `{id}` part must be replaced by the identifier in decimal format given in the function call.

The resolved data of the URI must be a JSON blob.

## JSON scheme of Metadata

The JSON blob resolved through the uri must follow the following JSON schema.

If metadata were retrieved by a function call containing a token identifier and the string `{id}` exists in any JSON value, it MUST be replaced with the actual token id in decimal format, by all client software that follows this standard.

```
{
    "$schema": "http://json-schema.org/draft-07/schema#",
    "title": "Token Metadata",
    "type": "object",

>     "required": ["version", "name"],

Reviewer 4: isnt "image" required?

Author: I removed the requirement for image, because `properties.image_data` could be used or `image_data` directly. One of these three needs to be defined. I don't know how to specify that in the schema.

                }
            }
        },
        "image_data": {
            "type": "string",
            "description": "Raw SVG image data. Deprecated. Use `properties.image_data`."
        },
        "external_url": {
            "type": "string",
            "description": "Url to view the item on a 3rd party web site. Deprecated. Use `properties.external_url`."
        },
        "animation_url": {
            "type": "string",
            "description": "URL to a multi-media attachment for the item. Deprecated. Use `properties.animation_url`."
        },
    }
}
```

The lengths of string values is not restricted. Nowadays, clients should be smart enough to deal with values of different lengths.

### Example

token101.json

```
{
  "version": 1,
  "name": "Foo #101",
  "image": "ipfs://somerandomecid",
  "attributes": [
     {
      "trait_type": "hair",
      "value": "red",
    },
    {
      "trait_type": "strength",
      "display_type": "number",
      "value": 99,
    },
  ],
  "properties": {
      "collection":  "Foo Collection",

>       "collection_size":  "10000"

Reviewer 4: The example here uses "collection_size", but below, under common properties, "total_supply" is specified. I would like some clarity here. My expectation is that we will use "total_supply" for collection size.

Author: fixed

  },
  "localization": {
      "uri": "ipfs://somerandomcid/{locale}.json",
      "default": "en",
      "locales": ["en", "pt-BR", "de"]
  }
}
```

de.json

```
{
    "version": 1,

>     "attritbutes: [

Reviewer 2: another minor nit: typo here :)

Reviewer 2: and below

Author: fixed in 2bc81e987b43de79a4c91c16272947849e0d142c

| `category`                      | `string`  | category of the associated media file, e.g. `image`, `video`, `audio`, `vr`, `html`.                                                                                                                                                                                                                                                                                |
| `files`                         | `array`   | list of all associated files, represented as `{uri: string, type: string, signature: string, signature_type: string}`.                                                                                                                                                                                                                                              |

> | `creators`                      | `array`   | list of creators and their shares, represented as `{address: string, share: number}`.                                                                                                                                                                                                                                                                               |

Reviewer 4: How is "share" specified?  as a uint?  as a percentage?  Is the context for royalties?  Should royalties be speciifed separately?

Author: clarification has been added

```

### Properties
Common properties are described in Appendic C.

### Properties
Common properties are described in appendix C. Properties of type object are usually described using the following schema:
```
{
    "$schema": "http://json-schema.org/draft-07/schema#",
    "title": "Token Metadata Property",
    "type": "object",
    "required": ["value"],
    "properties": {
        "type": {
            "type": "string",
            "description": "type of the property"
        },
        "description": {
            "type": "string",
            "description": "description of the property"
        },
        "value": {

>             "type": any,

Reviewer 3: does `type: any` have complications if we wanted to bring this metadata into a contract?
esp given how strongly typed clarity is? you can't even cast a num to an int there.

Reviewer: Is `any` a valid type? https://stackoverflow.com/a/16826141/1141891

Author: Has been corrected

            "description": "value of the property"
        }
    }
}
```

Example:

```
{
    "type": "string", 
    "description": "Address of custodian key holder", 

>     "value": "Casa Inc., P.O. Box 20575, Charleston, S.C. 29413, United States."

Reviewer 3: good this is extensible this far, and not too complex... but i would rather focus on digital assets first.

Reviewer 4: In general disagree with focusing on digital assets only.

Author: This is only an example, even for digital assets there might be meta data that we don't want to specify here like `camera` or `iso`, ...
